tests/proofs/dexter/dexter-spec.md

```diff
@@ -122,6 +123,39 @@ If any of the conditions are not satisfied, the call fails.
       orBool IsUpdating
       orBool Sender =/=K CurrentManager
       orBool LQTAddress =/=K #Address("tz1Ke2h7sDdakHJQh8WX4Z372du1KChsksyU")
+      */
+```
+
+## Update Token Pool
+
+The contract queries its underlying token contract for its own token balance if the following conditions are satisfied:
+
+1.  the token pool is _not_ currently updating (i.e. `storage.selfIsUpdatingTokenPool = false`)
+2.  exactly 0 tez was transferred to this contract when it was invoked
+3.  the txn sender must be equal to the txn source
+4.  if we are running the FA2 version of Dexter, then check that the contract at address `storage.tokenAddress` has a well-typed FA2 `balance_of` entrypoint;
+    otherwise, check that the contract at address `storage.tokenAddress` has a well-typed FA12 `get_balance` entrypoint.
+
+```k
+  claim <k> #runProof(IsFA2, UpdateTokenPool) => . </k>
+        <stack> .Stack </stack>
+        <selfIsUpdatingTokenPool> false => true </selfIsUpdatingTokenPool>
+        <tokenAddress> TokenAddress:Address </tokenAddress>
+
+        <myaddr> SelfAddress </myaddr>
+        <myamount> #Mutez(Amount) </myamount>
+        <senderaddr> Sender </senderaddr>
+        <sourceaddr> Sender </sourceaddr>
+        <paramtype> #Type(nat) </paramtype>  // Why does this work????? Why would this be the type???
```

Comment:
This has to do with the fact that our semantics has never correctly handled a concept called `entrypoints`; unfortunately, this contract uses the entrypoints feature. What happens is: the contract has a certain parameter type.  Typically, when you invoke a contract, you pass in some value of the parameter type.

However, when the parameter type is a nested `or` type, we can label the leaves of the `or` with an entrypoint annotation. When you invoke a contract with an entrypoint, you don't need to mention the nested `or` part ---  you just need the entrypoint name and the leaf type.

In this case, the contract is looking up the type of its `update_token_pool_internal` entrypoint is --- (which is `nat`). However, the contract does not actually check that this entrypoint has any type in particular. It just fowards this information along to the contract that it later inovkes as a callback.

Reply:
So why would we need to have this type set when we are calling `update_token_pool`? I'd understand if we were calling `update_token_pool_internal`. Does this mean we can't handle more then one callback?